tests/pools/meta/integration/test_rate_handling.py

```diff
@@ -0,0 +1,234 @@
+import pytest
+from brownie import chain
+from brownie.test import strategy
+
+pytestmark = [
+    pytest.mark.usefixtures("add_initial_liquidity"),
+    # pytest.mark.skip_pool("template-meta"),
+]
+
+ETH_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"
+
+
+class StateMachine:
+    """
+    Stateful test that performs a series of deposits, swaps and withdrawals
+    on a meta pool and confirms that the virtual price only goes up.
+    """
+
+    st_pct = strategy("decimal", min_value="0.5", max_value="1", places=2)
+    st_rates = strategy("decimal[8]", min_value="1.001", max_value="1.004", places=4, unique=True)
+
+    def __init__(
+        cls,
+        alice,
+        base_swap,
+        swap,
+        wrapped_coins,
+        wrapped_decimals,
+        underlying_coins,
+        underlying_decimals,
+    ):
+        cls.alice = alice
+        cls.swap = swap
+        cls.base_swap = base_swap
+        cls.coins = wrapped_coins
+        cls.decimals = wrapped_decimals
+        cls.underlying_coins = underlying_coins
+        cls.underlying_decimals = underlying_decimals
+        cls.n_coins = len(wrapped_coins)
+
+    def setup(self):
+        # reset the virtual price between each test run
+        self.virtual_price_base = self.base_swap.get_virtual_price()
+        self.virtual_price = self.swap.get_virtual_price()
+
+        # approve base pool for swaps
+        base_coins = self.underlying_coins[self.n_coins - 1 :]
+        for idx in range(len(base_coins)):
+            base_coins[idx].approve(self.base_swap, 2 ** 256 - 1, {"from": self.alice})
```

Comment:
Moving this to `__init__` would mean it gets snapshotted and only runs once, will speed up the test some.